cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClient.java

```diff
@@ -122,14 +127,24 @@ public void sendMessages(ProgramRunId programRunId,
       try (Reader reader = new InputStreamReader(urlConn.getInputStream(), StandardCharsets.UTF_8)) {
         ProgramRunInfo responseBody = GSON.fromJson(reader, ProgramRunInfo.class);
         if (responseBody.getProgramRunStatus() == ProgramRunStatus.STOPPING) {
-          // TODO - CDAP-18744 - terminate the program based on the program endtimestamp payload in response body
+          // this will trigger controller.stop() in DefaultRuntimeJob
+          future.complete(true);
         }
       }
     } finally {
       closeURLConnection(urlConn);
     }
   }
 
+  public void acceptStopProgramCommandAsync(Consumer<Boolean> consumer) {
+    future.thenAcceptAsync(consumer);
```

Comment:
provide an executor - new demon thread for async exec
provide a name for the thread